gapis/api/service.proto

```diff
@@ -238,3 +238,117 @@ message CubemapLevel {
 	image.Info negative_z = 5;
 	image.Info positive_z = 6;
 }
+
+message Metrics {
+  oneof metrics {
+    MemoryBreakdown memory_breakdown = 1;
+  }
+}
+
+// The description of the memory layout of the API state
+message MemoryBreakdown {
+  // The API ID used for this call.
+  path.API api = 1;
+  // The memory allocations that are currently in use.
+  repeated MemoryAllocation allocations = 2;
+  // The index to get the constants names for MemoryAllocation `flags`.
+  // A value of -1 indicates no flag names should be fetched.
+  int32 allocation_flags_index = 3;
+  // The index to get the constants names for SparseImageBlock `aspects`.
+  // A value of -1 indicates no flag names should be fetched
+  int32 image_aspects_flags_index = 4;
+}
+
+// A single memory allocation
+message MemoryAllocation {
+  // The device this allocation was made on
+  uint64 device = 1;
+  // The memory type this allocation is from
+  uint32 memory_type = 2;
+  // Flags describing properties of the allocation, e.g. whether it's device
+  // local, whether it's host visible, etc.
+  uint32 flags = 3;
+  // Constant index for looking up the flag values with the path.ConstantSet endpoint
+  uint32 constant_set = 4;
+  // The API specific ID for this allocation
+  uint64 handle = 5;
+  // The user-readable name for this allocation
+  string name = 6;
+  // This size of the memory allocated
+  uint64 size = 7;
+  // The regions of host memory this allocation has been mapped into
+  MemoryMapping mapping = 8;
+  // All buffers/images bound to this memory allocation
+  repeated MemoryBinding bindings = 9;
+}
+
+// A mapping from part of a memory allocation into host memory
+message MemoryMapping {
+  // The size in bytes of the mapping
+  uint64 size = 1;
+  // The offset of the mapping into the source allocation, in bytes
+  uint64 offset = 2;
+  // The address on the host that this memory is bound to
+  uint64 mapped_address = 3;
+}
+
+// A buffer/image that is bound to part of a memory allocation
+message MemoryBinding {
+  // The API specific id for this object
+  uint64 handle = 1;
+  // The user-readable name for this binding
+  string name = 2;
+
+  // The size of the binding, in bytes
+  uint64 size = 3;
+  // The offset into the bound allocation, in bytes
+  uint64 offset = 4;
+
+  // Which type of object is using this memory, and any extra associated data
+  // with that.
+  oneof type {
```

Comment:
I am not sure if we can do this in Proto, but for consistency (and forward compatibility) can we make a BufferBinding and ImageBinding message, even if they are empty.